pkg/controller/direct/gkehub/featuremembership_controller.go

```diff
@@ -0,0 +1,213 @@
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+	// TODO(ziyue): resolve the external ref
+	projectID := obj.Spec.ProjectRef.Name
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve projectID")
+	}
+	mID, err := fullyQualifiedNameForMembership(*obj)
+	if err != nil {
+		return nil, err
+	}
+	fID, err := fullyQualifiedNameForFeature(*obj)
+	if err != nil {
+		return nil, err
+	}
+	apiObj, err := featureMembershipSpecKRMtoMembershipFeatureSpecAPI(&obj.Spec)
+	if err != nil {
+		return nil, err
+	}
+	return &gkeHubAdapter{
+		membershipID:  mID,
+		featureID:     fID,
+		projectID:     "projects/" + projectID,
+		location:      obj.Spec.Location,
+		desired:       apiObj,
+		featureClient: projectsLocationsFeaturesService,
+	}, nil
+}
+
+func (a *gkeHubAdapter) Find(ctx context.Context) (bool, error) {
+	if a.membershipID == "" || a.featureID == "" {
+		return false, nil
+	}
+	feature, err := a.featureClient.Get(a.featureID).Context(ctx).Do()
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting feature %q: %w", a.featureID, err)
+	}
+	a.actual = feature
+	return true, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *gkeHubAdapter) Delete(ctx context.Context) (bool, error) {
+	// emptying the membershipspec is sufficient
+	a.desired = &featureapi.MembershipFeatureSpec{}
+	if _, err := a.patchMembershipSpec(ctx); err != nil {
+		return false, fmt.Errorf("deleting membershipspec for %s: %w", a.membershipID, err)
+	}
+	return true, nil
+}
+
+func (a *gkeHubAdapter) patchMembershipSpec(ctx context.Context) ([]byte, error) {
+	feature := a.actual
+	// only change the feature configuration for the associated membership
+	feature.MembershipSpecs[a.membershipID] = *a.desired
+	op, err := a.featureClient.Patch(a.featureID, feature).UpdateMask("membershipSpecs").Context(ctx).Do()
+	if err != nil {
+		return nil, err
+	}
+	return op.Response, nil
+}
+
+func (a *gkeHubAdapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating gkehubfeaturemembership", "obj", u)
+
+	_, err := a.patchMembershipSpec(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to patch the MembershipSpec; %w", err)
+	}
+	log.V(2).Info("successfully created gkehubfeaturemembership")
+	// no need to set the status from the api response for the  &krm.GKEHubFeatureMembershipStatus{} as the it only has generic status.
+	return nil
+}
+
+func (a *gkeHubAdapter) Update(ctx context.Context, u *unstructured.Unstructured) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating object", "u", u)
+	actual := a.actual.MembershipSpecs[a.membershipID]
+	if !reflect.DeepEqual(a.desired.Configmanagement, &actual.Configmanagement) || !reflect.DeepEqual(a.desired.Policycontroller, &actual.Policycontroller) || !reflect.DeepEqual(a.desired.Mesh, &actual.Mesh) {
```

Comment:
I think you get away with this here because there are no output fields, I've been exploring some tricks to make this comparison easy even when there are output fields.  (E.g. if we map _back_ to Spec, then by definition those are only the input fields so we can compare spec with reflect.DeepEqual; but the downside is then we have to map from proto -> spec; but we might want that for export anyway; etc etc)

Reply:
Added a comment to indicate no output fields in the API objects.